src/fsharp/pars.fsy

```diff
@@ -1910,9 +1911,12 @@ opt_typ:
    | /* EMPTY */ { None } 
    | COLON typ { Some $2 }
 
-   
 atomicPatternLongIdent:
-  | UNDERSCORE DOT pathOp { let (LongIdentWithDots(lid,dotms)) = $3 in (None,LongIdentWithDots(ident("_",rhs parseState 1)::lid, rhs parseState 2::dotms)) }
+  | UNDERSCORE DOT pathOp {
+    if not (parseState.LexBuffer.SupportsFeature LanguageFeature.SingleUnderscorePattern) then
+        raiseParseErrorAt (rhs parseState 2) (FSComp.SR.parsUnexpectedSymbolDot())
```

Comment:
I'm not sure this approach will work in all features. Here effectively there was an error when this rule was absent, but what about features where the absence of the rule doesn't result in an error message but a different parser result?

Moreover I think the spirit of this switch is to provide a mechanism for breaking changes, when a feature resulted in an error (like the underscore feature) normally is not a breaking change.

I'm not sure what's the best way to solve this, maybe adding a `when feature` to the rule, but I think the fsy file doesn't support when clauses (though not sure).

Reply:
@gusty,

The current parser does not really allow for retry on pattern matches or filters, once it arrives in one of these blocks of code, the parsed source code has been successfully parsed.  Once we are here we pretty much can do only one of two things …  error out or modify the AST produced within this block.

If the source is to have a changed behavior based on language version, then produce an AST that reflects the needs of the new language feature, or if the new feature is not supported, then produce the AST required by the old language version.  If the source didn't match before then provoke an error.

---

The switch is not about allowing breaking changes in the language.  It's purpose is to allow users of the compiler to access a specific version of the language.

We expect source code that compiled and ran using the F# X version of the language to continue to successfully compile and run with F# X+1 or F# X+2 or F# X+99 of the language.

Some times we will fix bugs in the implementation of the tooling that will cause the above to not be achieved.  Cases where we fixed a bug should not be impacted by this switch.  On the other hand, often those bugs, become an intrinsic part of the language and never get fixed.

I hope this helps

Kevin

Reply:
Thanks for the explanation. If that's the case, it looks good to me.